Uses name and type attributes to declare interface vs gateway endpoints

lorengordon opened the pull request:
EDIT: 
AWS recently released s3 *interface* endpoints, in addition to the
prior s3 *gateway* endpoints. Unfortunately, the service name for
each is exactly the same, which makes it impossible in a single
attribute to determine which type the user wanted.

This patch works around that by using name and type attributes for
each service, so the user can specify exactly the type of endpoint
they want.

ORIGINAL:
AWS recently released s3 *interface* endpoints, in addition to the
prior s3 *gateway* endpoints. Unfortunately, the service name for
each is exactly the same, which makes it impossible in a single
attribute to determine which type the user wanted.

This patch works around that by exposting separate variables for
interface services and gateway services.

lorengordon commented:
@elementalvoid Sorry to ping you, but other than ourselves you're the only other contributor! Just wanted to give you a heads up about a backwards-breaking change in the works... Shouldn't be any tfstate changes required, just changes to input variables. Or if you have other ideas about how to support the new s3 *interface* endpoint, I'm all ears.

elementalvoid commented:
Hey! Thanks for the ping @lorengordon. We were just looking at this today. @songawee is on my team and was tackling it for us.

You took a different tact than we were thinking. We were planning on an upgrade to the 3.x AWS provider. But I do like that your update is just some input var changes and we don't thave to upgrade the provider.

Internally we may still push forward with the provider update as it's something we need to pursue anyway.

FWIW, with the 3.x provider we can fix the issue _and_ require zero changes to the API. Here's @songawee's 3.x provider working branch/pr on our fork. https://github.com/Ibotta/terraform-aws-tardigrade-vpc-endpoints/pull/3/files (note, our fork is a bunch of commits behind but that'll be easy to resolve).

In any case, we'll peak at this tomorrow for sure.

elementalvoid commented:
Hrm, another thought on changing the API/input variables: doing so makes this forward compatible for when AWS does this with Dynamo.

lorengordon commented:
Yeah, and this will also allow a user to choose *either* the s3/dynamodb gateway OR interface endpoint.... At a quick glance, I think @songawee's patch will always do both... And I'd need to test to see what happens when `data.aws_vpc_endpoint_service` returns zero results. That's *usually* an error condition for terraform data sources...

The other approach I was considering, was to change `var.vpc_endpoint_services` from `list(string)` to `list(object({...})` where each object must specify the service name and type. I think that approach and what I have here *currently* will be pretty much equivalent, in the end.

lorengordon commented:
@elementalvoid @songawee I went ahead and implemented the other approach also, so you can compare.

* Vars per endpoint type: [`2de7917` (#131)](https://github.com/plus3it/terraform-aws-tardigrade-vpc-endpoints/pull/131/commits/2de79173d5bf5d86fcffcd8300855241c9ed99e2)
* Object-typed var with embedded type info: https://github.com/plus3it/terraform-aws-tardigrade-vpc-endpoints/pull/131/files

Lemme know if you have a preference!

elementalvoid commented on variables.tf:
```diff
@@ -53,20 +53,21 @@ variable "sg_ingress_rules" {
-  description = "List of AWS Endpoint service names that are used to create VPC Interface Endpoints. Both Gateway and Interface Endpoints are supported. See https://docs.aws.amazon.com/general/latest/gr/rande.html for full list."
-  default     = []
+  description = "List of AWS Endpoint service names and types. Both Gateway and Interface Endpoints are supported. See https://docs.aws.amazon.com/general/latest/gr/rande.html for full list."
+  type = list(object({
```
🤔 I think a `list of maps` makes things terribly verbose/redundant as compared to the two lists.

I do generally like data structures as input vars though. What if the variable is defined as:
```hcl
variable "vpc_endpoint_services" {
  type = object({
    interface = list(string)
    gateway   = list(string)
  })
  default = {
    interface = []
    gateway   = []
  }
}
```

Then the input from the user looks like:
```hcl
module "foo" {
  vpc_endpoint_services = {
    interface = ["sns", "sqs", "config"]
    gateway   = ["s3", "dynamo"]
  }
}
```

Note that I'm typing the object declaration from my memory of tf syntax .. it could be wrong .. or just plain not be possible. But I wanted to pose the idea.

lorengordon replied:
You're not wrong on the verbosity, but I was also thinking about using that data structure in the future to do other interesting things... e.g. Have an externally-managed list of security groups you want to use with just this one endpoint? Ok, expose `security_groups` as another attribute.

Also, once the "optional" feature for object attributes goes live, we can also easily implement defaults, to make things much less verbose.

elementalvoid replied:
To add some context to where my `verbose/redundant` comment comes from: our "default" input var declaration will go from `22` lines to `85` lines. It's a tad of a nitpick, but at the moment we're only enabling `20` out of a possible ~`75` endpoints! If we chose to enable all of them we'd have to declare a variable at near `300` lines!

lorengordon replied:
Is that all in a tfvars file you use with this module directly? Or do you have a .tf file where you can construct that value using a for loop? The interface *you* expose to this module doesn't have to change, necessarily...

lorengordon replied:
Just pushed an example of the new `optional` attribute feature, currently only available as an experiment. But this is what I'm thinking to reduce a lot of that verbosity...

elementalvoid replied:
Thanks for the updates!

Our implementation is indeed behind an interface. We use this module in a couple of different ways and neither requires a user to "configure" it. One location wraps this in another module strictly for sharing a common set of enabled endpoints -- include the wrapper, no config passed, done. The second location is more opinionated about its endpoints but hides away the details from the user (automated, declarative, internal Python tool). So, in neither case is the verbosity a huge issue for us.

At the moment it still feels a tad on the edge of adding complexity for little to no benefit. But! The idea of being able to add additional config per-endpoint-service is interesting and requires such an object interface to make it not-ugly. I'm onboard I think.

---

A side note on the optional attribute: It _requires_ a tf 0.14 upgrade. That shouldn't be a big issue for anyone already on the version requiring 0.12. Though it will likely come as a surprise for many. The bigger question on my mind is if the _experimental_ feature should be used in a production module? I've been leary of Terraform experiments in the past. I think I am in favor of verbosity over experimental features. 😀 

If the `type` attribute is not optional you can stick with a minimum tf version of 0.12 but it'll be compatible all the way to 0.14. This is, to me, yet another win for verbosity.

lorengordon replied:
> The bigger question on my mind is if the experimental feature should be used in a production module?

I wouldn't implement the `optional()` modifier until that feature is released GA. It was just meant as an example of what it _could_ look like, in a future version.

> I'm onboard I think

I was chatting with another dev internally, and we're both favoring this typed object. So I'll clean up this branch and we can get this released!

elementalvoid replied:
Sounds good! Thanks for the considerate back and forth; it's much appreciated.

elementalvoid commented on versions.tf:
```diff
@@ -1,3 +1,5 @@
 terraform {
-  required_version = ">= 0.12.9"
+  required_version = ">= 0.14"
```
Since the new `service_type` argument of the `aws_vpc_endpoint_service` resource requires the 3.x AWS Provider it would be good to add (assuming tf 0.14 isn't adopted):

```hcl
terraform {
  required_version = ">= 0.12.26"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.0"
    }
  }
}
```
`0.12.26` added [limited support for the "required_providers" syntax](https://www.terraform.io/docs/language/providers/requirements.html#v0-12-compatible-provider-requirements).

This will make it so that anyone attempting to utilize the module on a stack locked to a 2.x provider will get an error message about incompatible provider versions.

lorengordon replied:
Reading that link:

> If the provider belongs to the hashicorp namespace, as with the hashicorp/aws provider shown above, omit the source argument and allow Terraform v0.13 to select the hashicorp namespace by default.

So I think the block should be:

```
terraform {
  required_version = ">= 0.12.26"

  required_providers {
    aws = {
      version = ">= 3.0"
    }
  }
}
```

lorengordon commented:
Eef, and in testing discovered that S3 Interface endpoints do not support private DNS. Gonna hard-code that exception for now... But that could be exposed as a new attribute with this typed-object approach...

lorengordon commented:
Added s3 interface creation to a couple tests, and fixed a bug enabling Private DNS for s3 interfaces.

Now, results of `make test`:
```
PASS
ok      tardigarde-ci/tests     1657.845s
[terratest/test]: Completed successfully!
```
🥳 🍾

elementalvoid reviewed (approved):
Looks good to me. Thanks for keeping us in mind as you developed the changes!

mcarmean-novetta reviewed (approved):
Looks good.

lorengordon commented:
@elementalvoid This is released as [tag 6.0.0](https://github.com/plus3it/terraform-aws-tardigrade-vpc-endpoints/releases/tag/6.0.0)!
